### PR TITLE
Search/migrate search results

### LINKED ISSUE #22847: Migrate the SearchResults component to React
**What would you like to be added**:
The first part of the Search migration to React. 

The goal is to migrate [SearchResults](https://github.com/grafana/grafana/blob/master/public/app/core/components/search/search_results.ts) component and try to use it from Angular.

### PR DESCRIPTION
**Pull request by Clarity-89**
**What this PR does / why we need it**:

First step for moving search components to React. 
- Refactor Search Results into React component and split it into smaller pieces. 
- Use the new components in dashboard search and the main search
- Add some initial component testing (will be expanded when a more clear state management is introduced)
- Fix `TagFilter` bug where removing a sole tag would cause an error.
- Small tweaks to `Tag` component to forward onClick event.

**Which issue(s) this PR fixes**:

Closes #22847 

**Special notes for your reviewer**:
- There shouldn't be any apparent UI changes except for the style of new checkboxes (the manage dashboards action section still uses the old checkbox styles until it's refactored). 
- Only the search results has been refactored. The next step would be to refactor the actions section from the manage dashboards search #23057. 
- Some temporary state management functionality (marked with TODOs) has been added to angular components to delay using React state management. I think it would make sense to delay introducing a unified state management on React side (e.g. Redux) until more components have been refactored and the overall structure is a bit more clear.

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
Took this branch for a spin, and read through the code again, looks good!

File: public/app/features/search/components/SearchItem.tsx
Comment (Clarity-89) [inline public/app/features/search/components/SearchItem.tsx]: Not sure if this and the bg gradient below should be moved to themes.

File: public/app/features/search/components/SearchResults.tsx
Comment (Clarity-89) [inline public/app/features/search/components/SearchResults.tsx]: Not sure if this should be moved to a separate file, since it's only used in `SearchResults` for now.

**Comment by Clarity-89** [comment]
@torkelo thanks for checking this out! The PR is now ready for review, let me know if I missed any functionality. 

Not 100% happy with the new components yet, there's quite a bit of prop passing and maybe some rendering performance could be improved (not sure how many dashboards are normally rendered), but I hope to address those issues when introducing some sort of state.

**Comment by torkelo** [comment]
If you want to test performance you can provision a ton of dashboards using:

https://github.com/grafana/grafana/blob/master/devenv/setup.sh#L68 

But I also recommend: 
npx grafana-toolkit searchTestData --count=1000 

You might have to update your admin password in https://github.com/grafana/grafana/blob/master/packages/grafana-toolkit/src/cli/tasks/searchTestDataSetup.ts#L12 

Remember to backup your grafana.db if you want to go back to few dashboards again :)

File: public/app/features/search/components/SearchItem.tsx
Comment (torkelo) [reply public/app/features/search/components/SearchItem.tsx]: We have some emotion style mixins (pushed an update that uses that)

**Comment by torkelo** [comment]
pushed an update, strange commit as CircleCI (been testing a CI packages publish script, must have done something to git config )

**Comment by Clarity-89** [comment]
@torkelo seems like it expects cla signature from circle CI or w/e that commit was pushed with. Any idea how to fix that?

**Comment by torkelo** [comment]
We can ignore it

**Review by torkelo — APPROVED** [review]

**Comment by Clarity-89** [comment]
@torkelo I can't merge it before all the checks have passed.

**Comment by torkelo** [comment]
let me know when it's ready to merge and I can merge it  (even when checks have not passed)

File: packages/grafana-ui/src/themes/mixins.ts
Comment (dprokop) [inline packages/grafana-ui/src/themes/mixins.ts]: stylesFactory please:)

File: public/app/features/search/components/SearchCheckbox.tsx
Comment (dprokop) [inline public/app/features/search/components/SearchCheckbox.tsx]: stylesFactory please

File: public/app/features/search/components/SearchItem.tsx
Comment (dprokop) [inline public/app/features/search/components/SearchItem.tsx]: stylesFactory please

File: public/app/features/search/components/SearchResults.tsx
Comment (dprokop) [inline public/app/features/search/components/SearchResults.tsx]: and again, stylesFactory ;)

File: packages/grafana-ui/src/themes/mixins.ts
Comment (torkelo) [reply packages/grafana-ui/src/themes/mixins.ts]: this is used in a other styles, it returns a string :)

File: public/app/features/search/components/SearchResults.tsx
Comment (dprokop) [inline public/app/features/search/components/SearchResults.tsx]: and one more time :)

File: public/app/features/search/components/SearchResults.tsx
Comment (torkelo) [inline public/app/features/search/components/SearchResults.tsx]: what does this cx function do? `cx(css`some css`, 'pointer', { selected })`  I do not understand this, is there a pointer class? and passing an object with a boolean property as last argument what does that do?

File: packages/grafana-ui/src/themes/mixins.ts
Comment (Clarity-89) [reply packages/grafana-ui/src/themes/mixins.ts]: Nice, totally missed that. But I guess we shouldn't wrap the mixins in it, since those are used inside other styles, which in turn are wrapped in the factory.

File: public/app/features/search/types.ts
Comment (torkelo) [inline public/app/features/search/types.ts]: this looks the same as https://github.com/grafana/grafana/blob/master/public/app/types/search.ts#L6 would be good to try to combine them if possible

File: public/app/features/search/components/SearchResults.tsx
Comment (Clarity-89) [reply public/app/features/search/components/SearchResults.tsx]: Class `pointer` adds cursor pointer on hover, not sure if we have anything like that in the new styles, I also wanted to reuse the existing styles for now. 
`{ selected }` adds a selected class to the element if `selected` prop is true. It's done that way so  hover and selected could be combined because they both have the same styles.  
```  
&:hover,
&.selected {
  color: ${theme.colors.text};
}
```

File: public/app/features/search/types.ts
Comment (Clarity-89) [reply public/app/features/search/types.ts]: Hmm, I think it'd make sense to have all these types inside the `features/search`, but that would involve fixing all the imports/references to the previous types so maybe do that at a later point when we're starting to remove the old search components?

Comment (torkelo) [reply public/app/features/search/types.ts]: yes, sounds like a plan

**Comment by Clarity-89** [comment]
@torkelo this is ready to be merged unless you have some more comments :)

**Pull request merged**
